### PR TITLE
BMA2XX Sensor

### PR DESCRIPTION
**Pull request by joshgrob**
This is a pull request to add a generic BMA200 accelerometer series sensor.  It includes changes made after a review with Vipul from last week (12/6).  In the hw/sensors/bma2xx directory is a short readme on operation.  Thank you for the time, and will address any comments as soon as possible.

### CODE REVIEW

File: hw/bsp/bmd300eval/src/hal_bsp.c [hunk 1]
```diff
@@ -74,6 +75,14 @@ static const struct nrf52_hal_spi_cfg os_bsp_spi0s_cfg = {
 };
 #endif
 
+#if MYNEWT_VAL(I2C_0)
+static const struct nrf52_hal_i2c_cfg hal_i2c_cfg = {
+    .scl_pin = 15,
```
[the remark this answers is not in the document]
Comment (joshgrob) [reply hw/bsp/bmd300eval/src/hal_bsp.c]: Thank you for the comments Rymanluk.  I will make sure to add the topic prefix to commits.  There are three differences from studying the datasheets:
1.  Chip Id Value from register 0x00
2. G scale values as you mentioned above
3. Bandwidth settings.  On BMA280 there is no 1000hz option (just 500 and ODR max which is unfiltered 2000hz).

About hal_bsp.c - I agree that using mynewt_vals for I2C pins would make it consistent.  Will do make the change.

File: hw/bsp/bmd300eval/src/hal_bsp.c [hunk 2]
```diff
@@ -182,6 +191,11 @@ hal_bsp_init(void)
     assert(rc == 0);
 #endif
 
+#if MYNEWT_VAL(I2C_0)
+rc = hal_i2c_init(0, (void *)&hal_i2c_cfg);
```
[the remark this answers is not in the document]
Comment (joshgrob) [reply hw/bsp/bmd300eval/src/hal_bsp.c]: will do

File: hw/bsp/bmd300eval/syscfg.yml [hunk 3]
```diff
@@ -82,8 +92,10 @@ syscfg.vals:
     COREDUMP_FLASH_AREA: FLASH_AREA_IMAGE_1
     MCU_DCDC_ENABLED: 1
     XTAL_32768: 1
+    XTAL_32768_SYNTH: 0
```
[the remark this answers is not in the document]
Comment (joshgrob) [reply hw/bsp/bmd300eval/syscfg.yml]: agreed, will remove

File: hw/bsp/bmd300eval/syscfg.yml [hunk 4]
```diff
@@ -82,8 +92,10 @@ syscfg.vals:
...
+    XTAL_RC: 0
+    BLE_XTAL_SETTLE_TIME: 1500
 
 syscfg.vals.BLE_LP_CLOCK:
     OS_CPUTIME_FREQ: 32768
     OS_CPUTIME_TIMER_NUM: 5
-    BLE_XTAL_SETTLE_TIME: 1500
```
[the remark this answers is not in the document]
Comment (joshgrob) [reply hw/bsp/bmd300eval/syscfg.yml]: agreed.

File: hw/drivers/nimble/nrf52/src/ble_phy.c [hunk 5]
```diff
@@ -930,7 +930,6 @@ ble_phy_rx_start_isr(void)
     pdu_usecs = ble_phy_mode_pdu_start_off(ble_hdr->rxinfo.phy_mode) +
                 g_ble_phy_t_rxaddrdelay[ble_hdr->rxinfo.phy_mode];
     if (usecs < pdu_usecs) {
-        ticks--;
```
[the remark this answers is not in the document]
Comment (joshgrob) [reply hw/drivers/nimble/nrf52/src/ble_phy.c]: ok.

File: hw/drivers/sensors/bma2xx/pkg.yml [hunk 6]
```diff
@@ -28,6 +28,11 @@ pkg.keywords:
...
     - "@apache-mynewt-core/hw/hal"
+    - "@apache-mynewt-core/hw/sensor"
 
-pkg.deps.BMA253_LOG:
-    - "@apache-mynewt-core/sys/log/full"
+pkg.req_apis:
+    - stats
+    - log
+
+pkg.deps.BMA253_CLI:
```
[the remark this answers is not in the document]
Comment (joshgrob) [reply hw/drivers/sensors/bma2xx/pkg.yml]: good catch.  Thanks!

File: hw/sensor/include/sensor/sensor.h [hunk 7]
```diff
@@ -20,6 +20,8 @@
 #ifndef __SENSOR_H__
 #define __SENSOR_H__
 
+#include <string.h>
```
[the remark this answers is not in the document]
Comment (joshgrob) [reply hw/sensor/include/sensor/sensor.h]: ok

**Comment by joshgrob** [comment]
@vrahane I did do a rebase before submitting the fork.  In short I did a rebase from origin master (mynewt-core) to my local branch, then pushed that commit to my branch.  Is that not correct?

**Comment by vrahane** [comment]
That is correct, not sure how your history got screwed up. Did you do "git pull --rebase origin master". This would basically put all your changes/commits above other changes/commits.

File: hw/drivers/sensors/bma2xx/src/bma2xx.c [hunk 8]
```diff
@@ -0,0 +1,4872 @@
...
+#if MYNEWT_VAL(SPI_0_MASTER) || MYNEWT_VAL(SPI_1_MASTER)
+static struct hal_spi_settings spi_bma2xx_settings = {
+        .data_order = HAL_SPI_MSB_FIRST,
+        .data_mode  = HAL_SPI_MODE0,
+        .baudrate   = 4000,
+        .word_size  = HAL_SPI_WORD_SIZE_8BIT,
+};
+#endif
+
+enum int_route int_route;
```
[the remark this answers is not in the document]
Comment (joshgrob) [reply hw/drivers/sensors/bma2xx/src/bma2xx.c]: Sorry @rymanluk I can't tell which section you thought was not needed.  From what I see above I have not changed much from the BMA253 driver.  Lastly, should I change this value below? 

#define LOG_MODULE_BMA2XX (253)

File: hw/drivers/sensors/bma2xx/syscfg.yml [hunk 9]
```diff
@@ -18,6 +18,48 @@
...
+    BMA2XX_SPI_CLK:
+      description: 'Clock pin for SPI interface'
+      value:
+    BMA2XX_LOG:
+        description: 'Enable BMA2XX logging'
+        value: 0
+    BMA2XX_SHELL_DEV_NAME:
+        description: 'BMA2XX Shell device name'
+        value: "\"bma2xx_0\""
+    BMA280:
```
[the remark this answers is not in the document]
Comment (joshgrob) [reply hw/drivers/sensors/bma2xx/syscfg.yml]: That is odd, I thought I removed those BMA253, BMA280 sysconfig values below.  Right now they are not used, and there is a enum (enum bma2xx_model) in the struct bma2xx_cfg.model attribute that you set to indicate the specific sensor.  Will that work for you?

[the remark this answers is not in the document]
Comment (joshgrob) [reply hw/drivers/sensors/bma2xx/syscfg.yml]: I also agree with you about the config flags in the driver.  Unfortunately, it's not elegant in that it requires modification to sure the driver source which is not as intuitive as a config value.  Agreed to punt for now.

**Comment by joshgrob** [comment]
@vrahane I think I remember my issue.  I did git pull origin master first then a git pull --rebase origin master which I think jacked things up (you should be able to see the rebase merge commit message in the log).  Is that okay or should I do something else to fix things up?  One day I fully understand the mechanics of git :(

File: hw/drivers/sensors/bma2xx/src/bma2xx.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (rymanluk) [reply hw/drivers/sensors/bma2xx/src/bma2xx.c]: In github I cannot point to correct lines I'm commenting - sorry. 

"enum int_route int_route;" this is not needed. I believe it is copy-paste issue from bma253 which was fixed recently

File: hw/drivers/sensors/bma2xx/syscfg.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (rymanluk) [reply hw/drivers/sensors/bma2xx/syscfg.yml]: enum bma2xx_model is just perfect. Actually after I just open this PR to put comment  saying that we should have it configurable. Basically all the specific BMA2XXX stuff shall be pass as configuration. Thanks !  :+1:

[the remark this answers is not in the document]
Comment (joshgrob) [reply hw/drivers/sensors/bma2xx/syscfg.yml]: I think I see the confusion here.  In my original pr I didn't squash all commits so it should an updated commit with the BMA280 and BMA253 config values.  The latest commit has these removed

File: hw/sensor/include/sensor/sensor.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (joshgrob) [reply hw/sensor/include/sensor/sensor.h]: Hi @rymanluk on second thought can we keep this change in?  The reason is this breaks the build for my test application.  The sensor_test app compiles fine, but my app must have a different order of dependencies and breaks.  Would like to fix this in one shot.

**Comment by joshgrob** [comment]
@rymanluk I am closing this pull request since it getting confusing with the out dated commits and rebasing will open a new PR referencing the changes here.

File: hw/sensor/include/sensor/sensor.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (rymanluk) [reply hw/sensor/include/sensor/sensor.h]: I've create one for you: https://github.com/apache/mynewt-core/pull/703

**Pull request closed without merging**